elements/lisk-utils/src/math/q.ts

```diff
@@ -0,0 +1,153 @@
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		BigInt(intStr) * denominator +
+		(BigInt(fractionalStr) * denominator) / BigInt(10) ** BigInt(fractionalStr.length)
+	);
+};
+
+export const q = (val: number | bigint | Buffer, base: number | bigint): Q =>
+	Q.fromValue(val, base);
+
+export class Q {
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._base = BigInt(base);
+		if (this._base > MAX_FRAC) {
+			throw new Error('Base exceeds max fraction allowed.');
+		}
+		this._val = val;
+		this._base = BigInt(base);
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
+		if (val instanceof Q) {
+			return new Q(val._val, val._base);
+		}
+		const bigintBase = BigInt(base);
+		if (typeof val === 'number') {
+			return new Q(numberToQ(bigintBase, val), bigintBase);
+		}
+		if (typeof val === 'bigint') {
+			return new Q(val * TWO ** bigintBase, bigintBase);
+		}
+		return new Q(BigInt(`0x${val.toString('hex')}`), bigintBase);
+	}
+
+	public add(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Base of input does not match.');
+		}
+		return new Q(this._val + n._val, this._base);
+	}
+
+	public sub(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		if (this._val < n._val) {
+			throw new Error('Invalid input, output cannot be negative.');
+		}
+		return new Q(this._val - n._val, this._base);
+	}
+
+	public mul(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		return new Q((this._val * n._val) >> this._base, this._base);
+	}
+
+	public div(n: Q) {
+		if (this._base !== n._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		return new Q((this._val << this._base) / n._val, this._base);
+	}
+
+	public muldiv(n: Q, m: Q) {
+		if (this._base !== n._base || this._base !== m._base) {
+			throw new Error('Invalid input, base does not match.');
+		}
+		const x = this._val * n._val;
+		const y = x << this._base;
+		const z = y / m._val;
+
+		return new Q(this._roundDown(z), this._base);
+	}
+
+	public inv() {
+		const numerator = new Q(ONE << this._base, this._base);
+		return numerator.div(this);
+	}
+
+	public toInt(operation: Q_OPERATION = Q_OPERATION.ROUND_DOWN) {
+		if (operation === Q_OPERATION.ROUND_DOWN) {
+			return this._roundDown(this._val);
+		}
+		return this._roundUp(this._val);
+	}
+
+	public toBuffer(): Buffer {
+		return Buffer.from(this._val.toString(16), 'hex');
+	}
+
+	public eq(n: Q): boolean {
+		return this._val === n._val;
+	}
+
+	public lt(n: Q): boolean {
+		return this._val < n._val;
+	}
+
+	public lte(n: Q): boolean {
+		return this._val <= n._val;
+	}
+
+	public gt(n: Q): boolean {
+		return this._val > n._val;
+	}
+
+	public gte(n: Q): boolean {
+		return this._val >= n._val;
+	}
+
+	private _roundDown(n: bigint) {
```

Comment:
Rename to `floor()` and make public?

Reply:
Remove the internal usage and changed it to floor

Reply:
Actually the usage is not same as `floor`, internally so ill keep this and expose ceil/floor using those function